conan/tools/files/conandata.py

```diff
@@ -31,3 +31,31 @@ def recursive_dict_update(d, u):
     recursive_dict_update(conandata, data)
     new_content = yaml.safe_dump(conandata)
     save(path, new_content)
+
+
+def trim_conandata(conanfile):
+    """
+    Tool to modify the ``conandata.yml`` once it is exported, to limit it to the current version
+    only
+    """
+    if not hasattr(conanfile, "export_folder") or conanfile.export_folder is None:
+        raise ConanException("The 'trim_conandata()' can only be used in the 'export()' method")
+    path = os.path.join(conanfile.export_folder, "conandata.yml")
+    if not os.path.exists(path):
+        raise ConanException("conandata.yml file doesn't exist")
+
+    conandata = load(path)
+    conandata = yaml.safe_load(conandata)
+
+    version = str(conanfile.version)
+    result = {}
+    for k, v in conandata.items():
+        if not isinstance(v, dict):
+            result[k] = v
+            continue  # to allow user extra conandata, common to all versions
+        version_data = v.get(version)
+        if version_data is not None:
+            result[k] = {version: version_data}
+
+    new_conandata_yml = yaml.safe_dump(result, default_flow_style=False)
+    save(path, new_conandata_yml)
```

Comment:
I think it's easy to lose all the local modifications this way without any ability to restore.
at minimum, it should create a backup of an old conandata file.
but ultimately, I would prefer if it accepted `conandata_in` and `conandata_out` as arguments, so recipe immutability principle is not violated here.

Reply:
I don't know what you are talking about.
It is not possible to create any other file than the stripped/trimmed ``conandata.yml``, because that would break the immutability of recipe revisions, which is exactly the goal of this feature.
This is not new functionality, but moving what the hook is already doing to making it explicit in the recipes (it is biting too many users that are forking conan-center-index)

Reply:
Hi @SSE4 thanks for your input, I just wanted to clarify one point:

> I think it's easy to lose all the local modifications this way without any ability to restore.

and
> in this PR conandata.yml is modified in place

are not "right". See the PR linked test, where it shows how this is going to be used in the export method (Way before any manifest is created), and that it runs in the export folder, so no actual user file is overwritten, just the one that gets _exported_, so the user won't see any difference :)

Reply:
I am afraid this is not how it works, and https://github.com/conan-io/tribe/pull/16 is not related at all.
Whatever executes in ``export()`` happens **before** the recipe is fully created in the cache, **before** the ``conanmanifest.txt`` is computed and **before** the recipe revision is computed. It cannot break immutability because there is still nothing to be immutable, and the ``conandata.yml`` hasn't been stored yet.

Reply:
@RubenRBS 
> not right
if you take a look at changes:
```
    conandata = load(path)
    conandata = yaml.safe_load(conandata)
    ...
    save(path, new_conandata_yml)
```
it reads and write file from the same path. it means the file (`conandata.yml`) is modified in place.
I am not talking about manifest (it's irrelevant here), I don't see how manifest is connected at all.
what I mean is there is a file in conan cache (in `export_folder` specifically, which is obviously a part of a conan cache) named `conandata.yml` and it's modified inside the cache (as code does load/save on a file within cache).
so what I mean, is `conandata.yml`is stored in cache, and then `trim_conandata` modifies the file that is already stored in cache prior to the execution of the [export](https://docs.conan.io/2/reference/conanfile/methods/export.html?highlight=export_folder) method.
that violates tribe/16 proposal as it clearly says `Packages and recipes, identified by their complete reference including the revision, all artifact bare file contents (conanfile.py, source files, compiled libraries and executables, etc) stored in the Conan cache and in servers are always immutable.`.
it reads as once a file (e.g. `conandata.yml`) is stored within cache (e.g. `export_folder`) it's immutable. it cannot be mutated (modified) after it was already stored.
clearly, no file modification can happen on a file stored in `export_folder`.
`The immutability applies to the actual file contents: the bytes stored in the different raw files: conanfile.py, C++ sources and headers, compiled libraries (.lib, .a, .so), executables (.exe), etc.` - it means byte contents of `conandata.yml` cannot change.
so my suggestion is to keep immutability and do something like:
```
   conandata = load(path)
   conandata = yaml.safe_load(conandata)
   ...
   self.new_conandata = tempfile.NamedTemporaryFile().name
   save(self.new_conandata, new_conandata_yml)
```
and then just read from `self.new_conandata` (e.g. from [apply_conandata_patches](https://docs.conan.io/2/reference/tools/files/patches.html?highlight=apply%20conandata#conan.tools.files.patches.apply_conandata_patches) or whenever else needed).

> and the conandata.yml hasn't been stored yet.

it's stored in `self.export_folder` (which is in conan cache) at the beginning of `trim_conandata` call, and `load` reads stored file (otherwise, it wouldn't work here and raised for "file doesn't exist" - it uses `path = os.path.join(conanfile.export_folder, "conandata.yml")`).

Reply:
> that violates tribe/16 proposal as it clearly says 

No, it clearly doesnt. You just cited: ``Packages and recipes, identified by their complete reference including the revision, ``

The revision cannot be computed until the ``conanmanifest.txt`` is computed. And that is only computed after ``export()`` is executed. So the ``conanmanifest.txt`` is completely relevant here. 

It doesn't mean that every folder in the cache cannot be modified, otherwise it would be impossible to apply patches or build things, because the cache folders contents cannot be modified